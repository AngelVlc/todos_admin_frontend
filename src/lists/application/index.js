export * from "./GetListByIdUseCase";
export * from "./DeleteListByIdUseCase";
export * from "./GetListsUseCase";
export * from "./CreateListUseCase";
export * from "./UpdateListUseCase";
export * from "./MoveListItemUseCase";
export * from "./IndexAllListsUseCase";
export * from "./GetSearchSecureKeyUseCase";
