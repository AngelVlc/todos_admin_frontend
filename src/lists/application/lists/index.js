export * from "./GetListByIdUseCase";
export * from "./DeleteListByIdUseCase";
