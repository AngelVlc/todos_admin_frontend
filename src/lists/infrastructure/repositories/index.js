export * from "./ListsRepository";
export * from "./ListItemsRepository";
