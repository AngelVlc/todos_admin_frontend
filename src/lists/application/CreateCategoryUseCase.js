import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { CategoriesRepository } from "../infrastructure/repositories";

export class CreateCategoryUseCase extends BaseUseCase {
  _categoriesRepository;

  constructor({ categoriesRepository }) {
    super();

    this._categoriesRepository = categoriesRepository;
  }

  static create() {
    const categoriesRepository = new CategoriesRepository();

    return new CreateCategoryUseCase({ categoriesRepository });
  }

  async execute(category) {
   return await this._categoriesRepository.create(category);
  }
}
