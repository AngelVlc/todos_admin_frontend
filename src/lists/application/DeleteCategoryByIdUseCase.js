import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { CategoriesRepository } from "../infrastructure/repositories";

export class DeleteCategoryByIdUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new CategoriesRepository();

    return new DeleteCategoryByIdUseCase({ repository });
  }

  async execute(categoryId) {
    return await this._repository.deleteById(categoryId);
  }
}
