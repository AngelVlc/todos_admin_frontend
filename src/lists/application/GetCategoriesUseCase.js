import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { Category } from "../domain";
import { CategoriesRepository } from "../infrastructure/repositories";

export class GetCategoriesUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new CategoriesRepository();

    return new GetCategoriesUseCase({ repository });
  }

  async execute() {
    const result = await this._repository.getAll();

    return result.map((item) => {
      return new Category(item);
    });
  }
}
