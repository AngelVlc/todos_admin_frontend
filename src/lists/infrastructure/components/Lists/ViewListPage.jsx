import React, { useEffect, useState, useContext, useCallback } from "react";
import { useHistory, Link, useParams } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { GetListByIdUseCase } from "../../../application/lists";
import { GetListItemsUseCase } from "../../../application/listItems";
import { List } from "../../../domain";

export const ViewListPage = () => {
  let history = useHistory();
  let { listId } = useParams();
  const { useCaseFactory } = useContext(AppContext);
  const [pageState, setPageState] = useState(new List({ name: "", items: [] }));

  const getExistingList = useCallback(async () => {
    const getListByIdUseCase = useCaseFactory.get(GetListByIdUseCase);
    const listData = await getListByIdUseCase.execute(listId);

    const getListItemsUseCase = useCaseFactory.get(GetListItemsUseCase);
    const itemsData = await getListItemsUseCase.execute(listId);

    setPageState(new List({ ...listData, items: itemsData }));
  }, [listId, useCaseFactory]);

  useEffect(() => {
    getExistingList();
  }, [listId, getExistingList]);

  return (
    <div className="container">
      <h3 className="title">{`List '${pageState.name}'`}</h3>
      <nav className="breadcrumb" aria-label="breadcrumbs">
        <ul>
          <li>
            <Link to={`/`}>Home</Link>
          </li>
          <li>
            <Link to={`/lists`}>Lists</Link>
          </li>
          <li className="is-active">
            <Link aria-current="page" to={`/lists/${listId}`}>
              {pageState.name}
            </Link>
          </li>
        </ul>
      </nav>
      <div>
        {pageState.items.map((item) => (
          <div className="mb-2" key={item.id}>
            <p>{item.title}</p>
            <p className="is-size-7">{item.description}</p>
          </div>
        ))}
      </div>
      <div className="field is-grouped mt-4">
          <p className="control">
            <button className="button" data-testid="edit" onClick={() => history.push(`/lists/${listId}/edit`)}>
              EDIT
            </button>
          </p>
          <p className="control">
            <button
              className="button"
              data-testid="cancel"
              type="button"
              onClick={() => history.push("/lists")}
            >
              BACK
            </button>
          </p>
        </div>
    </div>
  );
};
