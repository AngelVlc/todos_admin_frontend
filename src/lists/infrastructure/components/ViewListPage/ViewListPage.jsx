import React, { useEffect, useState, useContext, useCallback } from "react";
import { useHistory, useParams } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { GetListByIdWithItemsUseCase } from "../../../application/lists";
import { Breadcrumb } from "../../../../shared/infrastructure/components/Breadcrumb";

export const ViewListPage = () => {
  let history = useHistory();
  let { listId } = useParams();
  const { useCaseFactory } = useContext(AppContext);
  const [pageState, setPageState] = useState();

  const getExistingList = useCallback(async () => {
    const getListByIdWithItemsUseCase = useCaseFactory.get(GetListByIdWithItemsUseCase);
    const list = await getListByIdWithItemsUseCase.execute(listId);

    setPageState(list);
  }, [listId, useCaseFactory]);

  useEffect(() => {
    getExistingList();
  }, [listId, getExistingList]);

  return (
    <>
      {pageState && (
        <div className="container">
          <Breadcrumb
            items={[
              { url: "/lists", text: "Lists" },
              { url: `/lists/${listId}`, text: pageState.name },
            ]}
          />
          <h3 className="title">{`List '${pageState.name}'`}</h3>
          <div>
            {pageState.items.map((item) => (
              <div className="mb-2" key={item.id}>
                <p>{item.title}</p>
                <p className="is-size-7">{item.description}</p>
              </div>
            ))}
          </div>
          <div className="field is-grouped mt-4">
            <div className="control">
              <button
                className="button"
                data-testid="edit"
                onClick={() => history.push(`/lists/${listId}`)}
              >
                EDIT
              </button>
            </div>
            <div className="control">
              <button
                className="button"
                data-testid="cancel"
                type="button"
                onClick={() => history.push("/lists")}
              >
                CANCEL
              </button>
            </div>
          </div>
        </div>
      )}
    </>
  );
};
