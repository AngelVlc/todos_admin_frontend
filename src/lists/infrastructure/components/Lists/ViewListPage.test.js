import { render, cleanup } from "@testing-library/react";
import { ViewListPage } from "./ViewListPage";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { MemoryRouter, Route } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { List, ListItem } from "../../../domain";
import { GetListByIdUseCase } from "../../../application/lists";

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const renderWithContextAndRouter = () => {
  const mockedGetListByIdUseCase = {
    execute: jest.fn(),
  };

  const mockedGetListItemsUseCase = {
    execute: jest.fn(),
  };

  const useCaseFactory = {
    get: (useCase) => {
      if (useCase == GetListByIdUseCase) {
        return mockedGetListByIdUseCase;
      }

      return mockedGetListItemsUseCase;
    },
  };

  mockedGetListByIdUseCase.execute.mockResolvedValue(
    new List({ id: 2, name: "list name" })
  );

  mockedGetListItemsUseCase.execute.mockResolvedValue([
    new ListItem({
      id: 5,
      title: "item title",
      description: "item description",
    }),
  ]);

  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <MemoryRouter initialEntries={[`/lists/2/read`]}>
          <Route path="/lists/:listId/read">
            <ViewListPage />
          </Route>
        </MemoryRouter>
      </AppContext.Provider>
    ),
  };
};

afterEach(cleanup);

it("should match the snapshot for an existing list", async () => {
  let fragment;
  await act(async () => {
    const { asFragment } = renderWithContextAndRouter();
    fragment = asFragment;
  });
  expect(fragment()).toMatchSnapshot();
});
