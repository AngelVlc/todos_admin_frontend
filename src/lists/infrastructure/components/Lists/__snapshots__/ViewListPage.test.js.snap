// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot for an existing list 1`] = `
<DocumentFragment>
  <div>
    <div
      class="container"
    >
      <nav
        aria-label="breadcrumbs"
        class="breadcrumb"
      >
        <ul>
          <li>
            <a
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              href="/lists"
            >
              Lists
            </a>
          </li>
          <li
            class="is-active"
          >
            <a
              aria-current="page"
              href="/lists/2"
            >
              list name
            </a>
          </li>
        </ul>
      </nav>
      <h3
        class="title"
      >
        List 'list name'
      </h3>
      <div>
        <div
          class="mb-2"
        >
          <p>
            item title
          </p>
          <p
            class="is-size-7"
          >
            item description
          </p>
        </div>
      </div>
      <div
        class="field is-grouped mt-4"
      >
        <div
          class="control"
        >
          <button
            class="button"
            data-testid="edit"
          >
            EDIT
          </button>
        </div>
        <div
          class="control"
        >
          <button
            class="button"
            data-testid="cancel"
            type="button"
          >
            CANCEL
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
